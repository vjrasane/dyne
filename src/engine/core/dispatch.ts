import { types as T, utils as U } from "dyne-commons";
import Effect from "../effects/effect";

export type Dispatch = (msg: any) => void;

export type Queues = {
  msg: any[];
  effect: Effect<any>[];
};

const dispatcher = <A>(queue: A[], queueable: A): void => {
  U.exists(queueable) && queue.push(queueable);
};

export const msgEventListener = (
  listener: T.Generator<object>,
  dispatch: Dispatch
): T.Sink => (...args: any[]): void => dispatch(listener(...args));

export const getDispatcher = () => {
  const queues: Queues = {
    msg: [],
    effect: []
  };

  const dispatch: Dispatch = (queueable: Effect<any> | object): void =>
    queueable instanceof Effect
      ? dispatcher(queues.effect, queueable)
      : dispatcher(queues.msg, queueable);

  return {
    queues,
    dispatch
  };
};
