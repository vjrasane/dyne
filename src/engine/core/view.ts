import { utils as U, types as T } from "dyne-commons";

export type View<M> = ((model: M) => T.VirtualDom) | T.VirtualDom;

export type Viewer<M> = (model: M) => T.VirtualDom;

export const getViewer = <M>(view: View<M>): Viewer<M> => {
  if (!U.exists(view)) {
    throw new Error(
      `Expected 'view' to be a function or element, was: '${typeof view}'`
    );
  }

  return U.isFunction(view) ? <Viewer<M>>view : (_: M) => <T.VirtualDom>view;
};
