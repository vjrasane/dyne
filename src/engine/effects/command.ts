import { utils as U } from "dyne-commons";
import { CmdFx, ProcessorF, PromiseableF, Processor, Failure } from "./types";
import Effect from "./effect";
import { Effectable } from "./return";
import { Dispatch } from "../core/dispatch";

export class Command<I, O> extends Effect<CmdFx<I>> {
  success: Processor<I, O>;
  failure: Failure<O>;

  constructor(cmd: CmdFx<I>, success?: Processor<I, O>, failure?: Failure<O>) {
    super(cmd);
    this.success = U.exists(success) ? success : U.identity;
    this.failure = U.exists(failure) ? failure : U.identity;
  }

  execute = (dispatch: Dispatch): Promise<void> =>
    processCommand(dispatch, this);
}

const processResult = (dispatch: Dispatch) => <I, O>(
  processor: Processor<I, O>,
  result: I
): void =>
  dispatch(
    U.isFunction(processor) ? (<ProcessorF<I, O>>processor)(result) : processor
  );

const processCommand = async <I, O>(
  dispatch: Dispatch,
  cmd: Command<I, O>
): Promise<void> => {
  const process: <I, O>(
    processor: Processor<I, O>,
    result: I
  ) => void = processResult(dispatch);

  try {
    const result = U.isFunction(cmd.effect)
      ? (<PromiseableF<I>>cmd.effect)()
      : cmd.effect;
    process(cmd.success, await result);
  } catch (error) {
    process(cmd.failure, error);
  }
};

export function Cmd<I, O>(
  cmd: CmdFx<I>,
  success?: Processor<I, O>,
  failure?: Failure<O>
): Command<I, O> {
  return new Command<I, O>(cmd, success, failure);
}

Cmd.wrap = <A>(wrappable: Effectable<A>): Effect<CmdFx<A>> => {
  if (wrappable instanceof Effect) {
    return wrappable;
  } else {
    return Cmd(wrappable);
  }
};
