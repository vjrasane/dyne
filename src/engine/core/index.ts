import { types as T } from "dyne-commons";
import { Setup, setup, Opts } from "./setup";
import { getDispatcher } from "./dispatch";

const UPDATE_INTERVAL = 1; /* milliseconds */
const VIEW_INTERVAL = 10; /* milliseconds */
const RENDER_INTERVAL = 10; /* milliseconds */
const EFFECT_INTERVAL = 1; /* milliseconds */

const intervals = {
  update: UPDATE_INTERVAL,
  view: VIEW_INTERVAL,
  render: RENDER_INTERVAL,
  effect: EFFECT_INTERVAL
};

export default <F, M>(setupObj: Setup<F, M>) => {
  const { dispatch, queues } = getDispatcher();
  const engine = setup(dispatch, queues, setupObj);
  return (opts: Opts<F>): void => {
    const procedures = engine(opts);
    Object.entries(procedures).forEach(([name, procedure]) =>
      setInterval(procedure, intervals[name])
    );
  };
};
