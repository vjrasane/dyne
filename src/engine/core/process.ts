import { utils as U } from "dyne-commons";
import Return from "../effects/return";
import { Updated, Update } from "./update";
import { Dispatch } from "./dispatch";

export default <A>(
  dispatch: Dispatch,
  updated: Updated<A>,
  previous?: A
): A => {
  const processed: Return<A> = Update.wrap(updated);
  processed.effects.forEach(e => dispatch(e));
  return U.exists(processed.model) ? processed.model : previous;
};
