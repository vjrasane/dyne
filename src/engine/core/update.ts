import { utils as U } from "dyne-commons";
import Return, { Effectable } from "../effects/return";
import Effect from "../effects/effect";
import { CmdFx } from "../effects/types";
import process from "./process";
import { Dispatch } from "./dispatch";

const throwUpdateError = (update: any): (() => never) =>
  U.throwError(
    new Error(
      `Attempted to pass message to 'update', but it was not a function, was: ${typeof update}`
    )
  );

export type Updated<A> = Return<A> | Effect<any> | CmdFx<A> | A;

export type Updater<A> = (msg: object, model: A) => A;

export type Update<A> = (msg: object, model: A) => Updated<A>;

export const Update = <A>(model: A, ...effects: Effectable<any>[]): Return<A> =>
  new Return(model, ...effects);

Update.wrap = <A>(wrappable: Updated<A>): Return<A> => {
  if (wrappable instanceof Return) {
    return wrappable;
  } else if (
    U.isFunction(wrappable) ||
    wrappable instanceof Promise ||
    wrappable instanceof Effect
  ) {
    return Update(null, wrappable);
  } else {
    return Update(<A>wrappable);
  }
};

export const getUpdater = <M>(
  dispatch: Dispatch,
  update: Update<M>
): Updater<M> => {
  if (U.isFunction(update)) {
    return (msg: object, model: M) =>
      process(dispatch, update(msg, model), model);
  }
  return throwUpdateError(update);
};
