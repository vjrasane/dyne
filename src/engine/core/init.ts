import { utils as U } from "dyne-commons";
import process from "./process";
import { Updated } from "./update";
import { Dispatch } from "./dispatch";

export type Init<F, M> = ((flags: F) => Updated<M>) | Updated<M>;

export type Initializer<F, M> = (flags: F) => M;

export const getInitializer = <F, M>(
  dispatch: Dispatch,
  init?: Init<F, M>
): Initializer<F, M> => {
  const initial: (flags: F) => Updated<M> = U.isFunction(init)
    ? <Initializer<F, M>>init
    : (_: F) => <Updated<M>>init;
  return (flags: F): M => process(dispatch, initial(flags));
};
