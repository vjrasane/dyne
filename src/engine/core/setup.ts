import { types as T, utils as U } from "dyne-commons";
import getRenderer, { Renderer } from "../render";
import { getInitializer, Initializer, Init } from "./init";
import { getUpdater, Updater, Update } from "./update";
import { getViewer, View } from "./view";
import { Dispatch, Queues } from "./dispatch";

export type Setup<F, M> = {
  init?: Init<F, M>;
  update?: Update<M>;
  view: View<M>;
};

export type Opts<F> = {
  node: HTMLElement;
  flags?: F;
};

export const setup = <F, M>(
  dispatch: Dispatch,
  queues: Queues,
  { init, update, view }: Setup<F, M>
) => {
  const initializer: Initializer<F, M> = getInitializer(dispatch, init);
  const updater: Updater<M> = getUpdater(dispatch, update);
  const viewer: View<M> = getViewer(view);
  const renderer: Renderer = getRenderer(dispatch);

  return ({ node, flags }: Opts<F>) => {
    let model: M = initializer(flags);

    const updateProcedure = () => {
      if (queues.msg.length > 0) {
        model = updater(queues.msg.shift(), model);
      }
    };

    let viewedModel: M;
    let virtualDom: T.VirtualDom;
    const viewVirtualDom = () => {
      virtualDom = viewer(model) || [];
      viewedModel = model;
    };

    const viewProcedure = () => {
      if (!virtualDom || (U.exists(model) && model !== viewedModel)) {
        viewVirtualDom();
      }
    };

    let renderedVirtualDom;
    const renderProcedure = () => {
      if (U.exists(virtualDom) && virtualDom !== renderedVirtualDom) {
        renderer(node, virtualDom, renderedVirtualDom);
        renderedVirtualDom = virtualDom;
      }
    };

    const effectProcedure = () => {
      if (queues.effect.length > 0) {
        const effect = queues.effect.shift();
        // execute effect with the message dispatcher
        return effect.execute(dispatch);
      }
    };

    return {
      update: updateProcedure,
      view: viewProcedure,
      render: renderProcedure,
      effect: effectProcedure
    };
  };
};
