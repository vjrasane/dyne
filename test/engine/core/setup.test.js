import Dyne from "../../../src";
import { setup } from "../../../src/engine/core/setup";
import { getDispatcher } from "../../../src/engine/core/dispatch";
import { Cmd } from "../../../src/engine/effects/command";
const html = `
<!DOCTYPE html>
<html>
  <head></head>
  <body>
    <div id="root"></div>
  </body>
</html>
`;
jest.useFakeTimers();

describe("setup", () => {
  let $root, dispatch, queues, setupBase;
  const clickHandler = event => ({ action: "event", event });
  const view = () => (
    <div id="clickable" onClick={clickHandler}>
      Hello World!
    </div>
  );
  beforeEach(() => {
    ({ dispatch, queues } = getDispatcher());
    setupBase = setupObj => setup(dispatch, queues, setupObj);
    document.body.innerHTML = html;
    $root = document.getElementById("root");
  });

  it("render with minimal setup", () => {
    const procedures = setupBase({ view })({ node: $root });

    procedures.view(); // initial dom
    procedures.render();

    expect($root.innerHTML).toMatchSnapshot();
  });

  it("does not render when dom has not changed", () => {
    const procedures = setupBase({ view })({ node: $root });

    procedures.view(); // initial dom
    procedures.render();
    procedures.render();

    expect($root.innerHTML).toMatchSnapshot();
  });

  it("does not create new dom when model has not changed", () => {
    const mockView = jest.fn().mockImplementation(view);
    const procedures = setupBase({ view: mockView })({ node: $root });
    procedures.view(); // initial dom
    mockView.mockClear();
    procedures.view();
    expect(mockView).not.toHaveBeenCalled();
  });

  it("does not update when no events are queued", () => {
    const update = jest.fn();
    const procedures = setupBase({ view, update })({ node: $root });
    procedures.update();
    expect(update).not.toHaveBeenCalled();
  });

  it("updates when events are queued", () => {
    const update = jest.fn();
    const procedures = setupBase({ view, update })({ node: $root });
    dispatch("message");
    procedures.update();
    expect(update).toHaveBeenCalledWith("message", undefined);
  });

  it("updates after click event", () => {
    const update = jest.fn();
    const procedures = setupBase({ view, update })({ node: $root });

    procedures.view(); // initial dom
    procedures.render();

    document
      .getElementById("clickable")
      .dispatchEvent(new window.Event("click"));

    procedures.update();

    expect(update.mock.calls).toMatchSnapshot();
  });

  it("creates new dom when model has changed", () => {
    const mockView = jest.fn().mockImplementation(view);
    const update = (msg, model) => ({ msg, model });
    const procedures = setupBase({ view: mockView, update })({ node: $root });
    dispatch("message");

    procedures.view(); // initial dom
    procedures.update();
    procedures.view();

    expect(mockView.mock.calls).toMatchSnapshot();
  });

  it("does not process effects if none are queued", () => {
    const mockDispatch = jest.fn();
    const procedures = setup(mockDispatch, queues, { view })({ node: $root });
    procedures.effect();
    expect(mockDispatch).not.toHaveBeenCalled();
  });

  it("processes queued effect", async () => {
    dispatch(Cmd("effect"));
    const mockDispatch = jest.fn();
    const procedures = setup(mockDispatch, queues, { view })({ node: $root });
    await procedures.effect();
    expect(mockDispatch).toHaveBeenCalledWith("effect");
  });
});
