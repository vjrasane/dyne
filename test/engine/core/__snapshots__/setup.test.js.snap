// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`setup creates new dom when model has changed 1`] = `
Array [
  Array [
    undefined,
  ],
  Array [
    Object {
      "model": undefined,
      "msg": "message",
    },
  ],
]
`;

exports[`setup does not render when dom has not changed 1`] = `"<div id=\\"clickable\\">Hello World!</div>"`;

exports[`setup render with minimal setup 1`] = `"<div id=\\"clickable\\">Hello World!</div>"`;

exports[`setup updates after click event 1`] = `
Array [
  Array [
    Object {
      "action": "event",
      "event": Event {
        "isTrusted": false,
      },
    },
    undefined,
  ],
]
`;
